File: SampleAppObjc/ViewController.m
//
//  ViewController.m
//  SampleAppObjc
//
//  Created by Markus Gasser on 26.11.15.
//  Copyright © 2015 konoma GmbH. All rights reserved.
//

#import "ViewController.h"
#import "ImageCell.h"

@import Matisse;



@interface ViewController ()

@property (nonatomic, readonly) CGFloat itemSize;
@property (nonatomic, readonly) NSArray<NSURL*> *imageURLs;

@end

@implementation ViewController

- (void)awakeFromNib {
    [super awakeFromNib];
    
    _itemSize = 150.0f;
    _imageURLs = @[
        [NSURL URLWithString:@"https://artseer.files.wordpress.com/2014/04/050rt_1.jpg"],
        [NSURL URLWithString:@"https://worldonaforkdotcom.files.wordpress.com/2013/10/untitled-126.jpg"],
        [NSURL URLWithString:@"https://bellaremyphotography.files.wordpress.com/2015/02/bma7feb15-01118.jpg"],
    ];
}

- (BOOL)prefersStatusBarHidden {
    return YES;
}

- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];
    
    CGFloat itemsPerRow = (CGFloat)floorf(self.view.bounds.size.width / self.itemSize);
    CGFloat exactItemSize = (CGFloat)floorf(self.view.bounds.size.width / itemsPerRow);
    
    ((UICollectionViewFlowLayout *)self.collectionViewLayout).itemSize = CGSizeMake(exactItemSize, exactItemSize);
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return 100;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    ImageCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"ImageCell" forIndexPath:indexPath];
    NSURL *url = self.imageURLs[indexPath.row % self.imageURLs.count];
    
    // Matisse.load(url).resizeTo(CGSize(width: itemSize, height: itemSize), contentMode: .ScaleAspectFill).into(cell.imageView)
    
    MTSMatisseContext *context = Matisse;
    
    [[[Matisse load:url]
      resizeTo:CGSizeMake(self.itemSize, self.itemSize) contentMode:UIViewContentModeScaleAspectFill]
     into:cell.imageView];
    
    NSLog(@"%@, %@", [context description], url);
    
    return cell;
}

@end

File: SampleApp/ViewController.swift
//
//  ViewController.swift
//  SampleApp
//
//  Created by Markus Gasser on 22.11.15.
//  Copyright © 2015 konoma GmbH. All rights reserved.
//

import UIKit
import Matisse


class ViewController: UICollectionViewController {
    
    private let itemSize: CGFloat = 150.0
    
    private let imageURLs = [
        NSURL(string: "https://artseer.files.wordpress.com/2014/04/050rt_1.jpg")!,
        NSURL(string: "https://worldonaforkdotcom.files.wordpress.com/2013/10/untitled-126.jpg")!,
        NSURL(string: "https://bellaremyphotography.files.wordpress.com/2015/02/bma7feb15-01118.jpg")!,
    ]
    
    override func prefersStatusBarHidden() -> Bool {
        return true
    }
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        
        let itemsPerRow = floor(view.bounds.width / itemSize)
        let exactItemSize = floor(view.bounds.width / itemsPerRow)
        
        (collectionViewLayout as! UICollectionViewFlowLayout).itemSize = CGSize(width: exactItemSize, height: exactItemSize)
    }
    
    override func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return 1000
    }
    
    override func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCellWithReuseIdentifier("ImageCell", forIndexPath: indexPath) as! ImageCell
        let url = imageURLs[indexPath.row % imageURLs.count]
        
        Matisse
            .load(url)
            .resizeTo(CGSize(width: itemSize, height: itemSize), contentMode: .ScaleAspectFill)
            .into(cell.imageView)
        
        return cell
    }
}

